parsers/CA_NB.py

```diff
@@ -0,0 +1,159 @@
+# The arrow library is used to handle datetimes consistently with other parsers
+import arrow
+
+# The request library is used to fetch content through HTTP
+import requests
+
+# BeautifulSoup is used to parse HTML to get information
+from bs4 import BeautifulSoup
+
+
+timezone = 'Canada/Atlantic'
+
+
+def _get_new_brunswick_flows(requests_obj):
+    """
+    Gets current electricity flows in and out of New Brunswick.
+
+    There is no reported data timestamp in the page. The page returns
+    current time and says "Times at which values are sampled may vary by
+    as much as 5 minutes."
+    """
+
+    url = 'https://tso.nbpower.com/Public/en/SystemInformation_realtime.asp'
+    response = requests_obj.get(url)
+
+    soup = BeautifulSoup(response.text, 'html.parser')
+
+    table = soup.find('table', attrs={'bordercolor': '#191970'})
+
+    rows = table.find_all('tr')
+
+    headers = rows[1].find_all('td')
+    values = rows[2].find_all('td')
+
+    flows = {headers[i].text.strip(): float(row.text.strip())
+             for i, row in enumerate(values)}
+
+    return flows
+
+
+def fetch_production(country_code='CA-NB', session=None):
+    """Requests the last known production mix (in MW) of a given country
+
+    Arguments:
+    country_code       -- ignored here, only information for CA-NB is returned
+    session (optional) -- request session passed in order to re-use an existing session
+
+    Return:
+    A dictionary in the form:
+    {
+      'countryCode': 'FR',
+      'datetime': '2017-01-01T00:00:00Z',
+      'production': {
+          'biomass': 0.0,
+          'coal': 0.0,
+          'gas': 0.0,
+          'hydro': 0.0,
+          'nuclear': null,
+          'oil': 0.0,
+          'solar': 0.0,
+          'wind': 0.0,
+          'geothermal': 0.0,
+          'unknown': 0.0
+      },
+      'storage': {
+          'hydro': -10.0,
+      },
+      'source': 'mysource.com'
+    }
+    """
+
+    """
+    In this case, we are calculating the amount of electricity generated
+    in New Brunswick, versus imported and exported elsewhere.
+    """
+
+    requests_obj = session or requests.session()
+    flows = _get_new_brunswick_flows(requests_obj)
+
+    # nb_flows['NB Demand'] is the use of electricity in NB
+    # 'EMEC', 'ISO-NE', 'MPS', 'NOVA SCOTIA', 'PEI', and 'QUEBEC'
+    # are exchanges - positive for exports, negative for imports
+    # Electricity generated in NB is then 'NB Demand' plus all the others
+
+    generated = (flows['NB Demand'] + flows['EMEC'] + flows['ISO-NE']
+                 + flows['MPS'] + flows['NOVA SCOTIA'] + flows['PEI']
+                 + flows['QUEBEC'])
+
+    data = {
+        'datetime': arrow.utcnow().datetime,
+        'countryCode': country_code,
+        'production': {
+            'unknown': generated
```

Comment:
Will it be used in calculations, so that PEI can know the estimated CO2/kWh they are importing from NB? (Using the default 700 value for 'unknown')